aten/src/ATen/native/vulkan/ops/Mm.cpp

```diff
@@ -7,94 +8,207 @@ namespace vulkan {
 namespace ops {
 namespace {
 
-Tensor addmm(
-    const Tensor& self_arg,
-    const Tensor& mat1_arg,
-    const Tensor& mat2_arg,
-    const Scalar beta,
-    const Scalar alpha) {
-  api::Context* const context = api::context();
+class LinearOpContext final : public torch::jit::CustomClassHolder {
+ public:
+  static LinearOpContext create(
+      api::Resource::Pool& pool,
+      const Tensor& weight,
+      const c10::optional<Tensor>& bias);
 
-  const Tensor self = self_arg.is_vulkan() ? self_arg : self_arg.vulkan();
-  const vTensor& v_self = convert(self);
+  using State = std::tuple<Tensor, c10::optional<Tensor>>;
 
-  const Tensor mat1 = mat1_arg.is_vulkan() ? mat1_arg : mat1_arg.vulkan();
-  const vTensor& v_mat1 = convert(mat1);
+  Tensor run(const Tensor& input, float beta, float alpha) const;
+  State unpack() const;
 
-  const Tensor mat2 = mat2_arg.is_vulkan() ? mat2_arg : mat2_arg.vulkan();
-  const vTensor& v_mat2 = convert(mat2);
+ private:
+  LinearOpContext(
+      api::Resource::Pool& pool,
+      const Tensor& weight,
+      const c10::optional<Tensor>& bias);
 
-  const auto self_sizes = self.sizes();
-  const auto mat1_sizes = mat1.sizes();
-  const auto mat2_sizes = mat2.sizes();
+ private:
+  struct {
+    vTensor v_weight;
+    vTensor v_bias;
+  } packed_;
+
+  struct {
+    Tensor weight;
+    c10::optional<Tensor> bias;
+  } unpacked_;
+};
+
+vTensor pack_weights(api::Resource::Pool& pool, const Tensor& weight_arg) {
+  if (weight_arg.is_vulkan()) {
+    return convert(weight_arg);
+  }
+  return convert(weight_arg.vulkan());
+}
 
-  if (self_sizes.size() >= 2) {
-    TORCH_CHECK(
-        (mat1_sizes[Layout::Parameter::width] ==
-            mat2_sizes[Layout::Parameter::height]) &&
-        (self_sizes[Layout::Parameter::height] ==
-            mat1_sizes[Layout::Parameter::height]) &&
-        (self_sizes[Layout::Parameter::width] ==
-            mat2_sizes[Layout::Parameter::width]),
-        "Incompatible matrix dimensions!");
+vTensor pack_biases(
+    api::Resource::Pool& pool,
+    const c10::optional<Tensor>& bias_arg,
+    const Tensor& weight_arg) {
+  if (bias_arg && bias_arg->is_vulkan()) {
+    return convert(*bias_arg);
   }
-  else {
-    TORCH_CHECK(
-        (mat1_sizes[Layout::Parameter::width] ==
-            mat2_sizes[Layout::Parameter::height]) &&
-        ((self_sizes[Layout::Parameter::height] ==
-            mat1_sizes[Layout::Parameter::height]) ||
-         (self_sizes[Layout::Parameter::height] ==
-            mat2_sizes[Layout::Parameter::width])),
-        "Incompatible matrix dimensions!");
+  if (bias_arg) {
```

Comment:
You can replace both ifs with the following:

```
if (bias_arg) {
  return convert(bias_arg->vulkan());
}

// The else logic goes here.  No need for else since you are returning from if.
```